### PR TITLE
Add side effect callbacks during retry

### PR DESCRIPTION
**Pull request by DybekK**
This PR addresses the proposed functionality in https://github.com/softwaremill/ox/issues/56 and introduces enhancements to the retry mechanism by adding a `onRetry` function callback to the `RetryPolicy` class. 

`onRetry` callback is executed after each retry attempt. It accepts two parameters: the attempt number and the result of the attempt. The result is represented as an Either type, where Left indicates an error and Right indicates a successful result. By default, this is an empty function.

### CODE REVIEW

**Review by kciesielski — COMMENTED** [review]
I don't think we need the `beforeEachAttempt` part, can't see any good use case for it. Users can simply enrich their retried effect with `effectBefore >> myEffect` and retry such composite effect. Adding this to the retry API could the flow a bit confusing. The `cats-retry` library offers a `onError: (E, RetryDetails) => M[Unit]` callback, and I think we should be fine with a similar thing. Wdyt @rucek @DybekK?

File: core/src/test/scala/ox/retry/RetryLifecycleTest.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    // then
+    result shouldBe successfulResult
+    counter shouldBe 1
+
+    beforeEachAttemptInvoked shouldBe true
+    beforeEachAttemptInvocationCount shouldBe 1
+
+    afterEachAttemptInvoked shouldBe true
+    afterEachAttemptInvocationCount shouldBe 1
+    returnedResult shouldBe Right(successfulResult)
```
Comment (rucek) [inline core/src/test/scala/ox/retry/RetryLifecycleTest.scala:51]: Aren't those assertions redundant? If you verify that the callback was invoked once, I guess one of them would be sufficient. Same for the "before" callback.

File: core/src/test/scala/ox/retry/RetryLifecycleTest.scala [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    // then
+    result shouldBe failedResult
+    counter shouldBe 4
+
+    beforeEachAttemptInvoked shouldBe true
+    beforeEachAttemptInvocationCount shouldBe 4
+
+    afterEachAttemptInvoked shouldBe true
+    afterEachAttemptInvocationCount shouldBe 4
+    returnedResult shouldBe Left(failedResult)
```
Comment (rucek) [inline core/src/test/scala/ox/retry/RetryLifecycleTest.scala:94]: Also here the assertions seem redundant - wouldn't it be sufficient to verify the invocation count here? Same for the "before" callback.

File: core/src/main/scala/ox/retry/retry.scala [hunk 3]
```diff
@@ -41,28 +52,36 @@ def retryEither[E, T](operation: => Either[E, T])(policy: RetryPolicy[E, T]): Ei
...
-def retry[E, F[_], T](em: ErrorMode[E, F])(operation: => F[T])(policy: RetryPolicy[E, T]): F[T] =
+def retry[E, F[_], T](em: ErrorMode[E, F])(operation: => F[T])(policy: RetryPolicy[E, T], lifecycle: RetryLifecycle[E, T]): F[T] =
   @tailrec
   def loop(attempt: Int, remainingAttempts: Option[Int], lastDelay: Option[FiniteDuration]): F[T] =
     def sleepIfNeeded =
       val delay = policy.schedule.nextDelay(attempt + 1, lastDelay).toMillis
       if (delay > 0) Thread.sleep(delay)
       delay
 
+    lifecycle.beforeEachAttempt(attempt + 1)
```
Comment (rucek) [inline core/src/main/scala/ox/retry/retry.scala:70]: I'm wondering if we could just number the attempts starting from 1 instead of 0 - WDYT?

**Review by rucek — COMMENTED** [review]
Some minor coments, but overall LGTM.

**Comment by rucek** [comment]
> Users can simply enrich their retried effect with `effectBefore >> myEffect` and retry such composite effect

@kciesielski the difefrence would be that `effectBefore` won't be aware of the attempt number, but `beforeEachAttempt` will. Not sure about an actual use case for that though.

**Comment by kciesielski** [comment]
Yes, I'm aware of that difference, but I can't see the usefulness of this attempt number in such a case. I think we should choose a simpler API and sacrifice it.

File: core/src/main/scala/ox/retry/retry.scala (hunk 3 above)
Comment (DybekK) [reply core/src/main/scala/ox/retry/retry.scala]: I agree. In my opinion, it is more intuitive to read the code if we get rid of `attempt + 1` in some places.

**Comment by DybekK** [comment]
> I don't think we need the `beforeEachAttempt` part, can't see any good use case for it. Users can simply enrich their retried effect with `effectBefore >> myEffect` and retry such composite effect. Adding this to the retry API could the flow a bit confusing. The `cats-retry` library offers a `onError: (E, RetryDetails) => M[Unit]` callback, and I think we should be fine with a similar thing. Wdyt @rucek @DybekK?

I completely removed `RetryLifecycle` class and replaced it with just `onRetry` function callback.

File: core/src/main/scala/ox/retry/retry.scala [hunk 4]
```diff
@@ -10,27 +10,49 @@ import scala.util.Try
...
   * @param policy
   *   The retry policy - see [[RetryPolicy]].
   * @return
   *   The result of the function if it eventually succeeds.
   * @throws anything
   *   The exception thrown by the last attempt if the policy decides to stop.
   */
-def retry[T](operation: => T)(policy: RetryPolicy[Throwable, T]): T =
-  retryEither(Try(operation).toEither)(policy).fold(throw _, identity)
+def retry[T](operation: => T, onRetry: OnRetry[Throwable, T] = emptyOnRetry)(policy: RetryPolicy[Throwable, T]): T =
```
Comment (adamw) [inline core/src/main/scala/ox/retry/retry.scala:22]: maybe `onRetry` should be then part of the `RetryPolicy`? Its docs says that `A policy that defines how to retry a failed operation.` - I think this easily includes an `onRetry` callback?

Comment (rucek) [reply core/src/main/scala/ox/retry/retry.scala:22]: I agree, I was thinking about that as well. Until we had two callbacks, it made sense to me to have them as a separate parameter, but as we have a single callback now, I'd make it a part of the `RetryPolicy` as well.

**Comment by adamw** [comment]
Thanks!

**Pull request merged**